File: app/payments.py
"""Real x402 payment verification for the MCP surface.

Why this file exists: the HTTP side is gated by the SDK's ASGI middleware, which
decodes the payment header, verifies it with the facilitator, runs the handler,
then settles. MCP has no middleware - the payment arrives as a tool argument. The
first version of this service gated MCP tools with `if not payment:`, which is a
presence check: `payment="x"` got the data for free. This module does the real
thing, through the SAME `x402ResourceServer` the middleware uses, so price, rails
and requirements cannot drift between the two surfaces.

Order of operations, deliberately matching the middleware:
    decode -> find matching requirements -> verify (facilitator)
    -> [caller runs rate limit / idempotency / the tool] -> settle (facilitator)

Verification and settlement are two steps for a reason: a caller who is over the
rate limit is refused *before* their money moves.
"""
from __future__ import annotations

import asyncio
import logging
from dataclasses import dataclass, field
from typing import Any

from .config import settings
from .x402_setup import get_resource_server, resource_configs

log = logging.getLogger("quanta.payments")

_initialized = False
# One lock per event loop: the app has one, the test suite spins up others, and
# an asyncio.Lock cannot be shared across loops.
_locks: dict[Any, asyncio.Lock] = {}


def _loop_lock() -> asyncio.Lock:
    loop = asyncio.get_running_loop()
    lock = _locks.get(loop)
    if lock is None:
        lock = _locks[loop] = asyncio.Lock()
    return lock


class FacilitatorUnavailable(RuntimeError):
    """The facilitator could not be reached, so we cannot price or verify."""


async def _ready() -> Any:
    """Return the initialized resource server, or None in unmetered dev mode.

    `initialize()` is a blocking HTTP call to the facilitator (it asks which
    scheme/network pairs it supports), so it runs in a thread and only once.
    """
    global _initialized
    server = get_resource_server()
    if server is None:
        return None
    if not _initialized:
        async with _loop_lock():
            if not _initialized:
                try:
                    await asyncio.to_thread(server.initialize)
                except Exception as exc:
                    raise FacilitatorUnavailable(str(exc)) from exc
                _initialized = True
    return server


def reset_for_tests() -> None:
    global _initialized
    _initialized = False


async def payment_requirements() -> list[Any]:
    """The accepted payment requirements, one per enabled rail, built by the SDK."""
    server = await _ready()
    if server is None:
        return []
    out: list[Any] = []
    for config in resource_configs():
        out.extend(server.build_payment_requirements(config))
    return out


async def mcp_challenge(route_key: str, resource: str = "") -> dict[str, Any]:
    """The x402 challenge an unpaid MCP tool call gets back.

    Same `accepts[]` the HTTP 402 carries, because it is built from the same
    server instance and the same ResourceConfig.
    """
    try:
        requirements = await payment_requirements()
    except FacilitatorUnavailable as exc:
        log.warning("cannot build challenge: facilitator unavailable (%s)", exc)
        return {"x402": "facilitator_unavailable", "reason": str(exc),
                "hint": "The payment facilitator is unreachable, so this tool cannot be priced right now."}
    return {
        "x402": "payment_required",
        "x402Version": 2,
        "resource": resource or route_key,
        "description": route_key,
        "accepts": [r.model_dump(by_alias=True, exclude_none=True) for r in requirements],
        "hint": ("Pay with an x402 client, then call this tool again with "
                 "payment=<base64 payment payload>."),
    }


@dataclass
class PaymentRefusal:
    """A payment that did not verify. The tool must not run."""

    reason: str
    message: str | None = None

    ok: bool = field(default=False, init=False)

    def as_dict(self) -> dict[str, Any]:
        out: dict[str, Any] = {"x402": "payment_invalid", "reason": self.reason}
        if self.message:
            out["message"] = self.message
        return out


@dataclass
class VerifiedPayment:
    """A payment the facilitator confirmed. Settlement has NOT happened yet."""

    payer: str
    network: str
    amount: str
    payload: Any
    requirements: Any
    server: Any

    ok: bool = field(default=True, init=False)
    tx_ref: str = field(default="", init=False)

    async def settle(self) -> tuple[bool, str]:
        """Move the money. Returns (success, tx_ref_or_error_reason)."""
        try:
            result = await self.server.settle_payment(self.payload, self.requirements)
        except Exception as exc:  # facilitator down mid-flight
            log.warning("settlement raised: %s", exc)
            return False, f"settle_error:{exc}"
        if not result.success:
            return False, result.error_reason or "settle_failed"
        self.tx_ref = result.transaction or ""
        if result.payer:
            self.payer = result.payer
        return True, self.tx_ref


async def verify_mcp_payment(
    payment_b64: str, route_key: str
) -> VerifiedPayment | PaymentRefusal:
    """Decode and verify a payment payload handed to an MCP tool.

    Returns VerifiedPayment (caller must call .settle() after the tool succeeds)
    or PaymentRefusal. Never raises on caller input.
    """
    from x402.http.utils import decode_payment_signature_header

    try:
        server = await _ready()
    except FacilitatorUnavailable as exc:
        return PaymentRefusal("facilitator_unavailable", str(exc))

    if server is None:
        # Unmetered dev mode: there is nothing to verify against. Refuse rather
        # than pretend a payment was checked.
        return PaymentRefusal("metering_disabled",
                              "This instance runs unmetered (X402_ENABLED=false); payments are not accepted.")

    # 1. decode - exactly the way the SDK middleware decodes the
    #    `payment-signature` / `x-payment` header.
    try:
        payload = decode_payment_signature_header(payment_b64)
    except Exception as exc:
        return PaymentRefusal("malformed_payment_payload", str(exc))

    # 2. does the payload match something we actually sell?
    try:
        available = await payment_requirements()
    except FacilitatorUnavailable as exc:
        return PaymentRefusal("facilitator_unavailable", str(exc))
    requirements = server.find_matching_requirements(available, payload)
    if requirements is None:
        return PaymentRefusal(
            "no_matching_requirements",
            f"payload does not match any accepted rail/price for {route_key}",
        )

    # 3. verify with the facilitator. Presence is not proof; this is.
    try:
        verified = await server.verify_payment(payload, requirements)
    except Exception as exc:
        log.warning("verify raised: %s", exc)
        return PaymentRefusal("verify_error", str(exc))
    if not verified.is_valid:
        return PaymentRefusal(verified.invalid_reason or "invalid_payment",
                              verified.invalid_message)

    payer = verified.payer or ""
    return VerifiedPayment(
        payer=payer,
        network=str(requirements.network),
        amount=settings.x402_price,
        payload=payload,
        requirements=requirements,
        server=server,
    )

File: src/payments.ts
/**
 * Real x402 payment verification for the MCP surface.
 *
 * Why this file exists: the HTTP side is gated by the SDK's Hono middleware,
 * which decodes the payment header, verifies it with the facilitator, runs the
 * handler, then settles. MCP has no middleware - the payment arrives as a tool
 * argument. The first version of this service gated MCP tools with
 * `if (!payment)`, which is a presence check: `payment="x"` got the data for
 * free. This module does the real thing, through the SAME x402ResourceServer the
 * middleware uses, so price, rails and requirements cannot drift between the two
 * surfaces.
 *
 * Order of operations, deliberately matching the middleware:
 *     decode -> find matching requirements -> verify (facilitator)
 *     -> [caller runs rate limit / the tool] -> settle (facilitator)
 *
 * Verification and settlement are two steps for a reason: a caller who is over
 * the rate limit is refused *before* their money moves.
 */

import { decodePaymentSignatureHeader } from "@x402/core/http";
import type { x402ResourceServer } from "@x402/core/server";
import type { PaymentPayload, PaymentRequirements } from "@x402/core/types";

import type { Settings } from "./config";
import { paymentRequirements, readyServer } from "./x402";

export interface PaymentRefusal {
  ok: false;
  reason: string;
  message?: string;
}

export interface VerifiedPayment {
  ok: true;
  payer: string;
  network: string;
  amount: string;
  nonce: string;
  settle: () => Promise<{ settled: boolean; txRef: string }>;
}

export type PaymentOutcome = VerifiedPayment | PaymentRefusal;

export function refusalBody(refusal: PaymentRefusal): Record<string, unknown> {
  const body: Record<string, unknown> = { x402: "payment_invalid", reason: refusal.reason };
  if (refusal.message) body.message = refusal.message;
  return body;
}

/**
 * Decode the payment header the middleware has ALREADY verified.
 *
 * The Hono middleware does not hand the verified payload down to the handler, so
 * the route decodes the same header again - only to read the payer address and
 * the nonce. This is not a second trust decision: by the time a handler runs, the
 * middleware has verified that header with the facilitator, and a request that
 * failed verification never reaches here.
 */
export function decodeVerifiedHeader(header: string | undefined | null): PaymentPayload | null {
  if (!header) return null;
  try {
    return decodePaymentSignatureHeader(header) as PaymentPayload;
  } catch {
    return null;
  }
}

/** The nonce is the only id the route and the settlement stamp can both see. */
export function nonceOf(payload: PaymentPayload | undefined | null): string {
  const inner = (payload as { payload?: Record<string, unknown> } | null)?.payload;
  const auth = (inner as { authorization?: Record<string, unknown> } | undefined)?.authorization;
  return String(auth?.nonce ?? "");
}

export function payerOf(payload: PaymentPayload | undefined | null): string {
  const inner = (payload as { payload?: Record<string, unknown> } | null)?.payload;
  const auth = (inner as { authorization?: Record<string, unknown> } | undefined)?.authorization;
  return String(auth?.from ?? "");
}

/**
 * The x402 challenge an unpaid MCP tool call gets back.
 *
 * Same `accepts[]` the HTTP 402 carries, because it is built from the same server
 * instance and the same ResourceConfig.
 */
export async function mcpChallenge(
  settings: Settings,
  description: string,
  resource: string,
): Promise<Record<string, unknown>> {
  let requirements: PaymentRequirements[];
  try {
    requirements = await paymentRequirements(settings);
  } catch (error) {
    return {
      x402: "facilitator_unavailable",
      reason: String((error as Error)?.message ?? error),
      hint: "The payment facilitator is unreachable, so this tool cannot be priced right now.",
    };
  }
  return {
    x402: "payment_required",
    x402Version: 2,
    resource,
    description,
    accepts: requirements,
    hint: "Pay with an x402 client, then call this tool again with payment=<base64 payment payload>.",
  };
}

/**
 * Decode and verify a payment payload handed to an MCP tool.
 *
 * Returns a VerifiedPayment (the caller must call `.settle()` after the tool
 * succeeds) or a PaymentRefusal. Never throws on caller input.
 */
export async function verifyMcpPayment(
  settings: Settings,
  paymentB64: string,
  routeKey: string,
): Promise<PaymentOutcome> {
  let server: x402ResourceServer | null;
  try {
    server = await readyServer(settings);
  } catch (error) {
    return { ok: false, reason: "facilitator_unavailable", message: String((error as Error)?.message ?? error) };
  }

  if (!server) {
    // Unmetered dev mode: there is nothing to verify against. Refuse rather than
    // pretend a payment was checked.
    return {
      ok: false,
      reason: "metering_disabled",
      message: "This instance runs unmetered (X402_ENABLED=false); payments are not accepted.",
    };
  }

  // 1. decode - exactly the way the SDK middleware decodes the PAYMENT-SIGNATURE header.
  let payload: PaymentPayload;
  try {
    payload = decodePaymentSignatureHeader(paymentB64) as PaymentPayload;
  } catch (error) {
    return { ok: false, reason: "malformed_payment_payload", message: String((error as Error)?.message ?? error) };
  }

  // 2. does the payload match something we actually sell?
  let available: PaymentRequirements[];
  try {
    available = await paymentRequirements(settings);
  } catch (error) {
    return { ok: false, reason: "facilitator_unavailable", message: String((error as Error)?.message ?? error) };
  }
  const requirements = server.findMatchingRequirements(available, payload);
  if (!requirements) {
    return {
      ok: false,
      reason: "no_matching_requirements",
      message: `payload does not match any accepted rail/price for ${routeKey}`,
    };
  }

  // 3. verify with the facilitator. Presence is not proof; this is.
  let verified: { isValid: boolean; invalidReason?: string; invalidMessage?: string; payer?: string };
  try {
    verified = (await server.verifyPayment(payload, requirements)) as typeof verified;
  } catch (error) {
    return { ok: false, reason: "verify_error", message: String((error as Error)?.message ?? error) };
  }
  if (!verified.isValid) {
    return {
      ok: false,
      reason: verified.invalidReason ?? "invalid_payment",
      ...(verified.invalidMessage ? { message: verified.invalidMessage } : {}),
    };
  }

  let payer = verified.payer ?? payerOf(payload);

  return {
    ok: true,
    payer,
    network: String(requirements.network),
    amount: settings.price,
    nonce: nonceOf(payload),
    settle: async () => {
      try {
        const result = await server.settlePayment(payload, requirements);
        if (!result.success) {
          return { settled: false, txRef: result.errorReason ?? "settle_failed" };
        }
        if (result.payer) payer = result.payer;
        return { settled: true, txRef: result.transaction ?? "" };
      } catch (error) {
        return { settled: false, txRef: `settle_error:${String((error as Error)?.message ?? error)}` };
      }
    },
  };
}
